apps/emqx_rule_engine/test/emqx_rule_engine_SUITE.erl

```diff
@@ -583,6 +584,94 @@ t_ensure_action_removed(_) ->
 %% Test cases for rule runtime
 %%------------------------------------------------------------------------------
 
+t_json_payload_decoding(_Config) ->
+    {ok, C} = emqtt:start_link(),
+    on_exit(fun() -> emqtt:stop(C) end),
+    {ok, _} = emqtt:connect(C),
+
+    Cases =
+        [
+            #{
+                select_fields =>
+                    <<"payload.measurement, payload.data_type, payload.value, payload.device_id">>,
+                payload => emqx_utils_json:encode(#{
+                    measurement => <<"temp">>,
+                    data_type => <<"FLOAT">>,
+                    value => <<"32.12">>,
+                    device_id => <<"devid">>
+                }),
+                expected => #{
+                    payload => #{
+                        <<"measurement">> => <<"temp">>,
+                        <<"data_type">> => <<"FLOAT">>,
+                        <<"value">> => <<"32.12">>,
+                        <<"device_id">> => <<"devid">>
+                    }
+                }
+            },
+            #{
+                select_fields => <<"payload as p, payload.f as p.answer">>,
+                payload => emqx_utils_json:encode(#{f => 42, keep => <<"that?">>}),
+                expected => #{
+                    <<"p">> => #{
+                        <<"f">> => 42,
+                        <<"keep">> => <<"that?">>,
+                        <<"answer">> => 42
+                    }
+                }
+            },
+            #{
+                select_fields => <<"payload as p, payload.f as p.jsonlike.f">>,
+                payload => emqx_utils_json:encode(#{
+                    jsonlike => <<"{\"a\":0000}">>,
+                    f => <<"huh">>
+                }),
+                %% behavior from 4.4: jsonlike gets wiped without preserving old "keys"
+                expected => #{
+                    <<"p">> => #{
+                        <<"jsonlike">> => #{<<"f">> => <<"huh">>},
```

Comment:
Hm, strange, I thought that bringing that clause back to `emqx_rule_maps` should make this testcase emit `<<"jsonlike">> => #{<<"a">> => 0, <<"f">> => <<"huh">>},`, now I'm wondering why it doesn't. 🤔

Ah, jiffy thinks it's invalid JSON. Maybe use `jsonlike => <<"{\"a\":0}">>` in the payload to better illustrate what's happening here?

Reply:
Hm true, I assumed that 0000 was a valid number in JSON :see_no_evil: :speak_no_evil:

Reply:
Actually, with that change, it diverges from the old e4.4 behavior: e4.4 drops the `a`, but the new version keeps it.